java/mapreduce/src/java/org/apache/orc/mapred/OrcList.java

```diff
@@ -71,4 +71,30 @@ public void readFields(DataInput input) throws IOException {
       }
     }
   }
+
+  @Override
+  public int compareTo(OrcList<E> other) {
+    if (other == null) {
+      return -1;
+    }
+    int ourSize = size();
+    int otherSize = other.size();
+    for(int e=0; e < ourSize && e < otherSize; ++e) {
+      Writable ours = get(e);
+      Writable theirs = other.get(e);
+      if (ours == null) {
+        if (theirs != null) {
+          return 1;
+        }
+      } else if (theirs == null) {
+        return -1;
+      } else {
+        int val = ((Comparable<Writable>) ours).compareTo(theirs);
```

Comment:
Comparing two complex types with different nested types will cause trouble here. To avoid this, we need to first compare the schemas, then the values. This applies to all the complex types.